clash-cores/src/Clash/Cores/Xilinx/Ethernet/Gmii/BlackBoxes.hs

```diff
@@ -0,0 +1,154 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# OPTIONS_HADDOCK hide #-}
+module Clash.Cores.Xilinx.Ethernet.Gmii.BlackBoxes where
+
+import Prelude
+
+import Clash.Annotations.Primitive
+import Clash.Backend (Backend)
+import Clash.Netlist.Types (BlackBoxContext(..), TemplateFunction(..))
+import Control.Monad.State (State)
+import Data.List.Infinite (Infinite(..), (...))
+import Data.String.Interpolate
+import Data.Text.Prettyprint.Doc.Extra (Doc)
+import Prettyprinter.Interpolate (__di)
+
+primName :: String
+primName = "Clash.Cores.Xilinx.Ethernet.Gmii.gmiiSgmiiBridgePrim"
```

Comment:
Noooo.... don't hardcode this please!

Reply:
I'm glad you split that out because it was a bit more hidden first and I had missed it in the first review.

Reply:
With the current module layout you could do this:
```diff
--- a/clash-cores/src/Clash/Cores/Xilinx/Ethernet/Gmii.hs
+++ b/clash-cores/src/Clash/Cores/Xilinx/Ethernet/Gmii.hs
@@ -148,4 +148,4 @@ gmiiSgmiiBridgePrim !_ !_  !_ !_ !_ !_ !_ !_ !_ !_ !_ !_ =
 
 {-# CLASH_OPAQUE gmiiSgmiiBridgePrim #-}
 {-# ANN gmiiSgmiiBridgePrim hasBlackBox #-}
-{-# ANN gmiiSgmiiBridgePrim (verilogPrim) #-}
+{-# ANN gmiiSgmiiBridgePrim (verilogPrim 'gmiiSgmiiBridgePrim) #-}
diff --git a/clash-cores/src/Clash/Cores/Xilinx/Ethernet/Gmii/BlackBoxes.hs b/clash-cores/src/Clash/Cores/Xilinx/Ethernet/Gmii/BlackBoxes.hs
index a493157f0..2b812f897 100644
--- a/clash-cores/src/Clash/Cores/Xilinx/Ethernet/Gmii/BlackBoxes.hs
+++ b/clash-cores/src/Clash/Cores/Xilinx/Ethernet/Gmii/BlackBoxes.hs
@@ -12,13 +12,11 @@ import Control.Monad.State (State)
 import Data.List.Infinite (Infinite(..), (...))
 import Data.String.Interpolate
 import Data.Text.Prettyprint.Doc.Extra (Doc)
+import Language.Haskell.TH (Name)
 import Prettyprinter.Interpolate (__di)
 
-primName :: String
-primName = "Clash.Cores.Xilinx.Ethernet.Gmii.gmiiSgmiiBridgePrim"
-
-verilogPrim :: Primitive
-verilogPrim =
+verilogPrim :: Name -> Primitive
+verilogPrim primName =
   let
     tfName = 'tclTemplateFunction
 
@@ -151,4 +149,4 @@ tclTemplate bbCtx
       }
     |]
   in pure bbText
-  | otherwise = error $ primName <> ": Expected exactly one component name"
+  | otherwise = error "gmiiSgmiiBridgePrim: Expected exactly one component name"
```

If you put the primitive in the same module as the black box, you could do this:
```haskell
{-# ANN gmiiSgmiiBridgePrim (
  let
    primName = 'gmiiSgmiiBridgePrim
    tfName = 'tclTemplateFunction
    ...
  in InlineYamlPrimitive [Verilog, SystemVerilog] [__i|
    BlackBox:
      name: #{primName}
      ...
```